lib/ABI/FunctionType.cpp

```diff
@@ -388,210 +417,447 @@ class ConversionHelper {
+  /// the c-style
+  ///
+  /// \param Used a set of registers confirmed to be in use by
+  ///        the function in question.
+  /// \param Allowed an ordered list of registers allowed to be used
+  ///        for returning values by the ABI.
+  /// \param PointerToCopyLocation an optional register used for pointing to
+  ///        return values in memory on some architectures. It is set to
+  ///        `model::Register::Invalid` if it's not applicable.
+  /// \param Binary a const reference to the binary containing information
+  ///        about already present types.
+  /// \param TypeDependencies a list of types that need to be added to the model
+  ///        for the returned return value to be valid.
+  ///
+  /// \tparam RegisterType type of a register, should be deduced from
+  ///         the \ref Used.
+  /// \tparam RegisterCount the number of registers for the current ABI, should
+  ///         be auto-magically deduced from the \ref Allowed.
+  ///
+  /// \return
   template<typename RegisterType, size_t RegisterCount, bool DryRun = false>
   static std::optional<model::QualifiedType>
-  convertReturnValue(const SortedVector<RegisterType> &UsedRegisters,
-                     const RegisterArray<RegisterCount> &AllowedRegisters,
-                     const model::Register::Values PointerToCopyLocation,
-                     model::Binary &TheBinary) {
-    if (UsedRegisters.size() == 0) {
-      auto Void = TheBinary.getPrimitiveType(model::PrimitiveTypeKind::Void, 0);
-      return model::QualifiedType{ Void, {} };
+  tryConvertingReturnValue(const SortedVector<RegisterType> &Used,
+                           const RegisterArray<RegisterCount> &Allowed,
+                           const model::Register::Values PointerToCopyLocation,
+                           const model::Binary &Binary,
+                           TypeVector &TypeDependencies) {
+    if (Used.size() == 0) {
+      // Short-circuit the functions that don't return anything (`void`).
+      return ensurePrimitiveTypeIsAvailable(model::PrimitiveTypeKind::Void,
+                                            0,
+                                            Binary,
+                                            TypeDependencies);
+    }
+
+    if (Allowed.empty()) {
+      // A return value is expected on an ABI that doesn't dedicate any
+      // registers to be used for those.
+      return std::nullopt;
     }
 
-    if (UsedRegisters.size() == 1) {
-      if (UsedRegisters.begin()->Location == PointerToCopyLocation) {
-        if constexpr (DryRun)
-          return model::QualifiedType{};
-        else
-          return getTypeOrDefault(UsedRegisters.begin()->Type,
-                                  PointerToCopyLocation,
-                                  TheBinary);
+    // Special case where the only used register is the pointer-to-copy location
+    if (Used.size() == 1 && Used.begin()->Location == PointerToCopyLocation) {
```

Comment:
We should check that the return type is the same as the first shadow argument and bail out otherwise. Shouldn't we?

Reply:
First of all, `PointerToCopyLocation` is not always the same as the first shadow argument. For example, `arm64` uses a separate register for it.

Also I'm a bit hesitant about asserting that the type is the same - consider that in a lot of cases we have no way of determining the type of the pointer in the aforementioned shadow-argument - we can't exactly expect an assertion that `pointer_or_number_64_t` points to a struct mentioned in the return value, can we?

Reply:
But wait, I'm now a lot more hesitant whether this `tryConvertingReturnValue` works as intended!
(read as I'm really surprised how I didn't run into any problems with that) - I'm checking the return value registers for this when I should actually be checking the arguments, since that's where the shadow argument would likely show up!

Reply:
> First of all, PointerToCopyLocation is not always the same as the first shadow argument. For example, arm64 uses a separate register for it

I mean same type, not same register.

> Also I'm a bit hesitant about asserting that the type is the same - consider that in a lot of cases we have no way of determining the type of the pointer in the aforementioned shadow-argument - we can't exactly expect an assertion that pointer_or_number_64_t points to a struct mentioned in the return value, can we?

I'm just suggesting to bail out if *the type* of the first shadow argument is not the same as *the type* of this return value.
No assertions.

Reply:
Ah wait, my bad. Both should in fact come from the very same analysis in a valid case. Yeah, I was overthinking it.

I totally agree.

Reply:
This code undertook quite a rework since then. I'll close this thread as irrelevant.

Feel free to re-open it (or create a new one) if you think the problem carries over to the new design as well.